src/java.base/share/classes/java/lang/AbstractStringBuilder.java

```diff
@@ -1702,145 +1854,214 @@ final byte getCoder() {
+     * @param index the index to insert the string
+     * @param str the string
+     */
+     private static byte[] putStringAt(byte[] value, byte coder, int count, int index, String str, int off, int end) {
+        byte[] newValue = inflateIfNeededFor(value, count, coder, str.coder());
+        coder = (newValue == value) ? coder : UTF16;
+        str.getBytes(newValue, off, index, coder, end - off);
+        return newValue;
+    }
+
+    /**
+     * {@return buffer with new characters appended, possibly inflated}
+     * The buffer may need to be inflated if any character does not fit;
+     * if the returned buffer is different then passed in, the new coder is UTF16.
+     * The caller is responsible for updating the count.
+     * @param value the current buffer
+     * @param coder the coder of the buffer
+     * @param count the character count
+     * @param s a char array
+     * @param off the offset of the first character to append
+     * @param end end last (exclusive) character to append
+     */
+    private static byte[] appendChars(byte[] value, byte coder, int count, char[] s, int off, int end) {
+        if (isLatin1(coder)) {
+            int latin1Len = StringUTF16.compress(s, off, value, count, end - off);
+            for (int i = off + latin1Len, j = count + latin1Len; i < end; i++) {
                 char c = s[i];
                 if (StringLatin1.canEncode(c)) {
-                    val[j++] = (byte)c;
+                    value[j++] = (byte)c;
                 } else {
-                    this.count = count = j;
-                    inflate();
-                    StringUTF16.putCharsSB(this.value, j, s, i, end);
-                    this.count = count + end - i;
-                    return;
+                    value = inflateToUTF16(value, j);
+                    // Store c to make sure sb has a UTF16 char
+                    StringUTF16.putCharSB(value, j++, c);
+                    i++;
+                    StringUTF16.putCharsSB(value, j, s, i, end);
+                    return value;
                 }
             }
         } else {
-            StringUTF16.putCharsSB(this.value, count, s, off, end);
+            StringUTF16.putCharsSB(value, count, s, off, end);
         }
-        this.count = count + end - off;
+        return value;
     }
 
-    private final void appendChars(String s, int off, int end) {
-        if (isLatin1()) {
+    /**
+     * {@return buffer with new characters appended, possibly inflated}
+     * The buffer may need to be inflated if any character does not fit;
+     * if the returned buffer is different then passed in, the new coder is UTF16.
+     * The caller is responsible for updating the count.
+     * @param value the current buffer
+     * @param coder the coder of the buffer
+     * @param count the character count
+     * @param s a string
+     * @param off the offset of the first character to append
+     * @param end end last (exclusive) character to append
```

Comment:
It might be helpful to somehow clarify that, in the case of appending Latin1 characters to a Latin1 buffer, it is expected/required that `value` is already of sufficient size.